src/os/bluestore/BlueFS.cc

```diff
@@ -3038,54 +3041,68 @@ void BlueFS::_consume_dirty(uint64_t seq)
   }
 }
 
-// Extends log if its free space is smaller then bluefs_min_log_runway.
-// Returns space available *BEFORE* adding new space. Signed for additional <0 detection.
-int64_t BlueFS::_maybe_extend_log()
-{
+int64_t BlueFS::_maybe_extend_log() {
+  uint64_t runway = log.writer->file->fnode.get_allocated() - log.writer->get_effective_write_pos();
```

Comment:
The overall idea of runway is to always have a available size of log in range of <min_log_runway, max_log_runway>.
That means, once we go lower than min_log_runway then we add so much that we should get above max_log_runway space.
It would be best if we could estimate how much free will be there *after* we write current transaction, and once we detect need for it, we will allocate new space, and add an op that extents log to *current* transaction. That way, we would never almost pay no cost for it.

In rare cases - which is the topic of this PR - it can happen that we run out of space for even current transaction. In such case we need to smartly inject log-extending transaction first, and then let normal processing take place.

Do you think it such a solution can be done?

Reply:
Thought it over and I have better idea.
I like your solution with always expanding log in separate transaction. It is simple and more controllable.
The cost is that we flush to disk and (potentially) wait to complete previous one.

Improvement proposal:
Do not flush ino.1 after log expanding transaction is written.
The data for the transaction will be in ino.1 write buffer, waiting for subsequent transaction to complete. It is not stable on disk, but we do not care, since we will need it after next transaction is written.